### PR TITLE
V0.4.0 update

### PR DESCRIPTION
**Pull request by AdamBuijks42**
Features: 
* Momentary assignments
* Tap-Tempo
* LED cycling for lists
* Static LED for lists blinks
* Bidirectional communication - Value set

### CODE REVIEW

File: src/hardware.h [hunk 1]
```diff
@@ -66,7 +65,8 @@ int hw_button(int button);
 void hw_led(int led, int color, int value);
 uint32_t hw_uptime(void);
 int hw_self_test(void);
-
+void hw_led_set(int led, int color, int value, int on_time_ms, int off_time_ms);
+int hw_settings_mode(void);
```
Comment (Janjhl) [inline src/hardware.h]: This function is never called, leftover code perhaps?

File: src/util.c [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+    // invert the string characters
+    str_len = (pstr - string);
+    reverse(string, str_len);
+
+    return str_len;
+}
+
+
+    while (1)
```
Comment (Janjhl) [inline src/util.c]: copy paste error?
This should not be here

File: src/util.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+/*
+****************************************************************************************************
+*       FUNCTION PROTOTYPES
+****************************************************************************************************
+*/
+
+uint32_t int_to_str(int32_t num, char *string, uint32_t string_size, uint8_t zero_leading, uint8_t need_minus);
+uint32_t float_to_str(float num, char *string, uint32_t string_size, uint8_t precision);
+
+void settings_screen_run(void);
```
Comment (Janjhl) [inline src/util.h]: This function is not used, old code?

File: state [hunk 4]
```diff
@@ -0,0 +1,12 @@
+src/hardware.c:84:            if (button->state == BUTTON_RELEASED)
```
Comment (Janjhl) [inline state]: accidental file creation?

File: Makefile [hunk 5]
```diff
@@ -16,13 +16,8 @@ CPU = LPC11U24
...
 
-# flags for debugging
-ifeq ($(DEBUG), 1)
-CFLAGS += -O0 -g -DDEBUG
-else
-CFLAGS += -O3
-endif
-
+CFLAGS += -Os
+
```
Comment (Janjhl) [inline Makefile:25]: let's leave the debug option, changing to Os is fine

**Review by Janjhl — CHANGES_REQUESTED** [review]
Small changes needed, as there is still some unused code
The rest seems ok!

**Comment by AdamBuijks42** [comment]
The requested changes should be fixed now.

File: src/util.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
+#ifndef SETTINGS_H
```
Comment (falkTX) [inline src/util.h]: this should be `UTILS_H`

**Review by Janjhl — APPROVED** [review]

**Pull request merged**